### PR TITLE
E2E: verify daemonset pods after machines

### LINKED ISSUE #2933: E2E tests aren't properly verifying the DaemonSets are running
/kind bug

[Before submitting an issue, have you checked the [Troubleshooting Guide](https://capz.sigs.k8s.io/topics/troubleshooting.html)?]

**What steps did you take and what happened:**
[A clear and concise description of what the bug is.]

When introducing the Calico HelmChart the Windows kube-proxy was moved into the calico namespace causing the upstream tests to fail. The fix for kube-proxy is in https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2931 but this issue is to track fixing the e2e framework to verify Daemonsets.

```
Warning  FailedCreate  8m49s (x18 over 19m)  daemonset-controller  Error creating: pods "kube-proxy-windows-" is forbidden: error looking up service account calico-system/kube-proxy: serviceaccount "kube-proxy" not found
```

This wasn't caught since the test doesn't really verify the number of pods running in the DS. This isn't isolated to Windows but also happens to Linux daemonsets too.

https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/c0f81b0b89e6640ccd6c9eee35aadac05ec36417/test/e2e/helpers.go#L276-L279

you can see this in the logs of the e2e tests, notice the log doesn't have the number of pods running

```
[1mSTEP[0m: waiting for daemonset calico-system/calico-node to be complete
Nov 29 21:42:44.456: INFO: waiting for daemonset calico-system/calico-node to be complete
Nov 29 21:42:44.489: INFO: daemonset calico-system/calico-node pods are running, took 33.187202ms
[1mSTEP[0m: Waiting for Ready calico windows pods
[1mSTEP[0m: waiting for daemonset calico-system/calico-node-windows to be complete
Nov 29 21:42:44.659: INFO: waiting for daemonset calico-system/calico-node-windows to be complete
Nov 29 21:42:44.691: INFO: daemonset calico-system/calico-node-windows pods are running, took 32.541702ms
[1mSTEP[0m: waiting for daemonset calico-system/kube-proxy-windows to be complete
Nov 29 21:42:44.831: INFO: waiting for daemonset calico-system/kube-proxy-windows to be complete
Nov 29 21:42:44.864: INFO: daemonset calico-system/kube-proxy-windows pods are running, took 32.690035ms
``` 

**What did you expect to happen:**
DaemonSets should be running before the e2e tests run.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]
Fixed the kubel

**Environment:**

- cluster-api-provider-azure version: 
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by jackfrancis**
**What type of PR is this?**

/kind failing-test

**What this PR does / why we need it**:

This PR moves the "validate expected daemonset pods" after "verify control plane is ready", because at that point in the cluster creation flow machines are not actually online, and so daemonset pods are never going to be scheduled.

Instead, we move daemonset pod validation after `ApplyClusterTemplateAndWait`, which waits for control plane *and* worker machines. At this point in the E2E flow we will have all of our expected number of machines verified as Ready, at which point we can check for the expected daemonset pods.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2933

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: test/e2e/azure_test.go [hunk 1]
```diff
@@ -466,6 +506,16 @@ var _ = Describe("Workload cluster creation", func() {
 				Args: []string{"--server-side"},
 			}, result)
 
+			By("Verifying that expected Daemonset Pods are running", func() {
+				EnsureDaemonsets(ctx, func() DaemonsetsSpecInput {
```
Comment (CecileRobertMichon) [inline test/e2e/azure_test.go:510]: are there any hooks in the e2e framework flow that can allow us to do it as part of create after the worker machines come up? Just like EnsureControlPlaneInitialized allows us to install the Helm charts after the control plane machines are up

File: test/e2e/azure_test.go [hunk 2]
```diff
@@ -186,6 +186,16 @@ var _ = Describe("Workload cluster creation", func() {
 					},
 				}, result)
 
+				By("Verifying that expected Daemonset Pods are running", func() {
```
Comment (marosset) [inline test/e2e/azure_test.go:189]: Can we do this with less code duplication?

**Comment by jackfrancis** [comment]
@CecileRobertMichon do you prefer this updated use of the `PostMachinesProvisioned` func?

@marosset this doesn't really address your valid observation. I'm not entirely certain why we wrap this test cases in a func and then invoke the input vars "just in time", rather than just passing them into the func as values. I assume there is some async value mutation that we need to properly track and that these func closures enable. Maybe @nojnhuh has more context?

File: test/e2e/daemonsets.go [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+// DaemonsetsSpecInput is the input for EnsureDaemonsets.
+type DaemonsetsSpecInput struct {
+	BootstrapClusterProxy framework.ClusterProxy
+	Namespace             *corev1.Namespace
+	ClusterName           string
+}
+
+// EnsureDaemonsets implements a test that verifies expected Daemonset Pods are running.
+func EnsureDaemonsets(ctx context.Context, inputGetter func() DaemonsetsSpecInput) {
```
Comment (CecileRobertMichon) [inline test/e2e/daemonsets.go:41]: why does this need to take an inputGetter as a func and not simply a `DaemonsetsSpecInput` as param?

**Comment by CecileRobertMichon** [comment]
> @CecileRobertMichon do you prefer this updated use of the PostMachinesProvisioned func?

yes 100%. Although not sure why we need to pass in a func as param, commented on that.

@marosset has a good point, there is a lot of duplication. Perhaps we could refactor our uses of the clusterctl `ApplyClusterTemplateAndWait` with a wrapper helper func that adds all the common stuff like this by default and just takes the stuff that changes as parameters. That's probably out of scope here but would be a nice code cleanup.

**Comment by jackfrancis** [comment]
Now that we're actually testing for the presence of the calico-node daemonset, it's failing three of the E2E test scenarios 🤯

File: test/e2e/azure_test.go (hunk 2 above)
Comment (marosset) [reply test/e2e/azure_test.go]: > > @CecileRobertMichon do you prefer this updated use of the PostMachinesProvisioned func?
> 
> yes 100%. Although not sure why we need to pass in a func as param, commented on that.
> 
> @marosset has a good point, there is a lot of duplication. Perhaps we could refactor our uses of the clusterctl `ApplyClusterTemplateAndWait` with a wrapper helper func that adds all the common stuff like this by default and just takes the stuff that changes as parameters. That's probably out of scope here but would be a nice code cleanup.

Maybe just leave a TODO comment?
I'm OK doing the cleanup later

Comment (CecileRobertMichon) [reply test/e2e/azure_test.go]: Let's open an issue (TODO comments tend to get lost)

**Comment by jackfrancis** [comment]
@jsturtevant @marosset see @CecileRobertMichon's comment above, I'm attempting to address that.

tl;dr 🤦, not 🤯.

**Comment by jackfrancis** [comment]
It definitely seems like our VMSS cluster template is failing to launch Windows csi-node-driver pods. See:

https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-azure/2950/pull-cluster-api-provider-azure-e2e/1611109298186752000/artifacts/clusters/capz-e2e-jfwxpp-vmss/kube-system/csi-azuredisk-node-win-6cc4v/azuredisk.log

@marosset @jsturtevant any ideas?

**Comment by CecileRobertMichon** [comment]
> It definitely seems like our VMSS cluster template is failing to launch Windows csi-node-driver pods. See:

That seems like the same error I was trying to fix in #2947

**Comment by CecileRobertMichon** [comment]
in theory rebase shouldn't solve it because prow should always rebase before running tests... so if it's still failing chances are #2947 didn't actually fix the issue

**Comment by jackfrancis** [comment]
@CecileRobertMichon It's still failing, and I'm noticing it on other PRs as well, is something environmental (different k8s version?) responsible?

**Comment by CecileRobertMichon** [comment]
looked at logs, it seems the template is missing the actual ClusterResourceSet for csi-proxy so even though the configmap is there it's not getting installed https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-azure/2950/pull-cluster-api-provider-azure-e2e/1611408924651556864/artifacts/clusters/bootstrap/capz-e2e-1ace1l-vmss-cluster-template.yaml

**Comment by jackfrancis** [comment]
@CecileRobertMichon indeed: https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2992

File: test/e2e/daemonsets.go (hunk 3 above)
Comment (jackfrancis) [reply test/e2e/daemonsets.go:41]: To answer this question definitively: I implemented this E2E helper func following the pattern of existing helper funcs.

Do either @nojnhuh or @mboersma know why this pattern exists?

**Comment by jackfrancis** [comment]
/retest

(cluster delete timeout flake)

**Comment by jackfrancis** [comment]
@CecileRobertMichon @marosset this is passing tests and ready for another review round

**Review by CecileRobertMichon — COMMENTED** [review]
This lgtm and is equivalent to previous functionality (but better because we're actually waiting for machines now). One thought: have you considered listing all daemonsets and waiting for all of them to to be available instead hardcoding a select few we can about?

If we want to make sure the daemonsets for calico exist we could potentially leave the existing code waiting for them to be available in the install Calico func and then do a general "wait for all daemonsets" post node provisioning

File: test/e2e/daemonsets.go (hunk 3 above)
Comment (jackfrancis) [reply test/e2e/daemonsets.go:41]: Summarizing a discussion: we concluded that this pattern is carried over from CAPI E2E; cycles permitting, we can investigate whether or not this is needed in all the places we've copied it in CAPZ E2E.

**Comment by jackfrancis** [comment]
@CecileRobertMichon this generalized approach appears to work fine, you can see the outcomes if you search for "Waiting for all DaemonSet Pods to be Running" in E2E output.

**Review by mboersma — COMMENTED** [review]
This looks good. Squash?

**Comment by jackfrancis** [comment]
@mboersma yeah just wanna get one more eyeball on this approach so I can revert if necessary

**Review by CecileRobertMichon — COMMENTED** [review]
lgtm

**Comment by CecileRobertMichon** [comment]
what's up with all the flakes 👀  anything related to this change?

**Comment by jackfrancis** [comment]
- make lint timeout
- default/web-windows7jo8s4 flake during LB IIS test
- HA test delete cluster timeout
- AKS E2E flake during autoscale test (addressed here: https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/3058)
- AKS E2E flake during public IP test (also addressed here: https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/3058)

I think that's it!

**Comment by jackfrancis** [comment]
@CecileRobertMichon sorry, not related to this change, I've seen every one of those flakes on other PRs :(
